back/engines/commercial/bulk_import_ideas/app/services/bulk_import_ideas/parsers/pdf/gpt_form_parser.rb

```diff
@@ -0,0 +1,136 @@
+    def parse_idea(file_path, page_count)
+      parsed_response = parse(file_path)
+
+      {
+        pdf_pages: (1..page_count).to_a,
+        fields: parsed_response&.map { |r| { r['text'] => r['answer'] } }&.reduce({}, :merge)
+      }
+    end
+
+    private
+
+    def parse(file_path)
+      pdf_file = Files::File.new(
+        name: 'scan.pdf',
+        mime_type: 'application/pdf',
+        content: File.open(file_path) # Uses the uploader
+      )
+
+      message = Analysis::LLM::Message.new(prompt, pdf_file)
+      gpt_response = llm.chat(message)
+
+      corrected_response = gpt_response.match(/\[.+\]/m)&.try(:[], 0) # to be sure it is json that can be parsed
+
+      corrected_response.present? ? JSON.parse(corrected_response) : nil
+      # binding.pry
+    end
+
+    def prompt
+      # In this message is a scanned survey form containing handwritten responses and checked options.
+      <<~GPT_PROMPT
+        In this message is a scanned survey form that has been filled in by hand.
+  
+        Your task is to extract the text and checked options based on the questions in the JSON form schema below.
+  
+        Do not attempt to interpret or summarize the responses. Ignore any text that does not appear handwritten. Do not fill in any missing answers.
+  
+        Do not extract answers unless they are in the correct order in the scanned form.
+
+        The language used in the form is #{@locale}.
+
+        Return the same array of JSON objects in the same order, but with an additional 'answer' attribute for each question, according to the following rules:
+        - if the question is a text question, add an attribute answer with the extracted handwritten text as value.
+        - if the question is of type select or multiselect then return an attribute answer as an array of the checked or ticked options' text values. If no options were checked, return an empty array.
+        - if the question is of type checkbox, return the answer as 'checked' if it has been checked.
+        - if the question is of type linear_scale, rating or sentiment_linear_scale, return the answer as a number.
+        - if the question is of type ranking, then there will be a number written in a box next to each option indicating its rank. Return the answer as an array of option texts ordered by their rank from lowest to highest. The written numbers will not be higher than the number of options.
+        - if the question is of type matrix_linear_scale, then there will checkboxes in rows against a list of matrix_statements on the left hand side. Return the answer as a hash with each statement as the key and the value as the number corresponding to the order (starting from 1, left to right) of the box that is checked or ticked for that row.
+        - if the question is optional, sometimes it may not be present in the scanned form.
+
+        Provide only the JSON without any additional text or markers.
```

Comment:
The recent SOTA models all support providing a JSON schema their answer should be compliant with. This is awesome, because we could not only encode these rules, we can already encode the very specific structure for this particular form, steering it exactly to what we need.

Just sharing this FYI, if this is already a solid improvement no need to block!